src/js/editor/post.js

```diff
@@ -798,13 +801,140 @@ class PostEditor {
     this.scheduleAfterRender(() => this.editor.selectRange(range));
   }
 
-  changeSectionTagName(section, newTagName) {
-    section.markers.forEach(m => {
-      m.clearMarkups();
-      this._markDirty(m);
+  /**
+   * @method toggleSection
+   *
+   * Toggles the active section or sections. If every section has the tag name, they will all be reset to default sections. Otherwise, every section will be changed to the requested type
+   *
+   * @param {String} sectionTagName A valid markup section or list section tag name (e.g. 'blockquote', 'h2', 'ul')
+   * @param {Range} range The range over which to toggle. Defaults to the current editor's offsets
+   * a list section
+   * @public
+   */
+  toggleSection(sectionTagName, range=this.editor.cursor.offsets) {
+    sectionTagName = normalizeTagName(sectionTagName);
+    let { post } = this.editor;
+    let everySectionHasTagName = true;
+    post.walkMarkerableSections(range, section => {
+      if (!this._isSameSectionType(section, sectionTagName)) {
+        everySectionHasTagName = false;
+      }
     });
-    section.tagName = newTagName;
-    this._markDirty(section);
+
+    let sectionsToChange = [];
+    post.walkMarkerableSections(range, section => {
+      sectionsToChange.push(section);
+    });
+    let tagName = everySectionHasTagName ? 'p' : sectionTagName;
+    sectionsToChange.forEach(section => {
+      this.changeSectionTagName(section, tagName);
```

Comment:
At one point there was an issue with mutation while walking — a section's `next` would get reassigned and the `walkMarkerableSections` callback would skip sections. I will try removing this and see if the tests still pass.